### DISCUSSION TITLE
PDF page rotation angle gives 90 degrees for visually straight readable page

### DISCUSSION THREAD

**Discussion by swathy-z9q** [pymupdf/PyMuPDF#1069] [Looking for help] [answered]
Hi,
I came across a pdf where visually the page is 0 degrees rotated, but the rotation attribute for the page gives a rotation angle of 90 degrees. How to get the actual rotation angle that is 0 degrees.

Please find the attached pdf replicating the scenario.
[pdf_rotation_anomaly.pdf](https://github.com/pymupdf/PyMuPDF/files/6582067/pdf_rotation_anomaly.pdf)

**Comment by JorjMcKie** [comment]
You are taking the wrong conclusion! Just because you can read the text like normal, does not mean that the page is unrotated!
Your "anomaly" is simply a decent, landscape page, on which text has been written from bottom-left to top-right, that's all.
Try it out yourself: set rotation to 0 and save the PDF with it.

**Reply by swathy-z9q** [reply]
Our use case is to find the visual rotation angle or the text direction. ( In this case, we expect 0 degrees).
When I tried setting the rotation angle 0 for the pdf, it is rotating the page back to -90 degrees. This is making the text direction readability from bottom-left to top-left.
I have manually created the attached(pdf_rotation_anomaly.pdf) pdf in my question to mimic the actual file we have got.
So is there any way we can figure out the visual rotated angle instead of rotation attribute?

**Comment by JorjMcKie** [comment]
Using PyMuPDF, you can easily create such an "anomaly" yourself!
[anomaly.zip](https://github.com/pymupdf/PyMuPDF/files/6583544/anomaly.zip)

**Comment by JorjMcKie** [comment, marked as answer]
> Our use case is to find the visual rotation angle or the text direction.

Let me strees again the point: it is no "anomaly" - but a very common thing.
As you already indicated, you need to take more than one information into account: the page rotation is one thing, another is the text orientation.
For the second aspect, you need to use a text extraction variant, which returns this information, too: `page.get_text("dict")`.
Its output is an object of nested dictionaries. Inside there you will find a dictionary for every text line. A line dictionary has the key "dir", which contains the requested information. `line["dir"]` is a 2-tuple of (cosine, sine) of the angle, which the line text has with the x-axis.
In the case of the example script's output PDF you will get this:
```python
>>> doc=fitz.open("rtext.pdf")
>>> page=doc[0]
>>> blocks = page.get_text("dict")["blocks"]
>>> for b in blocks:
	for line in b["lines"]:
		print(line["dir"])

		
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
(0.0, -1.0)
>>> 
```
The tuples (0, -1) mean that the angle has a cosine of zero and a sine value of -1. As expected: text goes from bottom to top.

**Reply by swathy-z9q** [reply]
Thank you so much. Your explanation helped me to get the required result.
